tailfs/compositefs/compositefs.go

```diff
@@ -0,0 +1,261 @@
+// sorted lexicographically by name.
+//
+// Children in a CompositeFileSystem can only be added or removed via calls to
+// the AddChild and RemoveChild methods, they cannot be added via operations
+// on the webdav.FileSystem interface like filesystem.Mkdir or filesystem.OpenFile.
+// In other words, the root of the CompositeFileSystem acts as read-only, not
+// permitting the addition, removal or renaming of folders.
+//
+// Rename is only supported within a single child. Renaming across children
+// is not supported, as it wouldn't be possible to perform it atomically.
+type CompositeFileSystem interface {
+	webdav.FileSystem
+	io.Closer
+
+	// AddChild ads a single child with the given name, replacing any existing
+	// child with the same name.
+	AddChild(child *Child)
+	// RemoveChild removes the child with the given name, if it exists.
+	RemoveChild(name string)
+	// SetChildren replaces the entire existing set of children with the given
+	// ones.
+	SetChildren(children ...*Child)
+	// GetChild returns the child with the given name and a boolean indicating
+	// whether or not it was found.
+	GetChild(name string) (webdav.FileSystem, bool)
+}
+
+// Options specifies options for configuring a CompositeFileSystem.
+type Options struct {
+	// Logf specifies a logging function to use
+	Logf logger.Logf
+	// StatChildren, if true, causes the CompositeFileSystem to stat its child
+	// folders when generating a root directory listing. This gives more
+	// accurate information but increases latency.
+	StatChildren bool
+	// Clock, if specified, determines the current time. If not specified, we
+	// default to time.Now().
+	Clock tstime.Clock
+}
+
+// New constructs a CompositeFileSystem that logs using the given logf.
+func New(opts Options) CompositeFileSystem {
+	logf := opts.Logf
+	if logf == nil {
+		logf = log.Printf
+	}
+	fs := &compositeFileSystem{
+		logf:         logf,
+		statChildren: opts.StatChildren,
+	}
+	if opts.Clock != nil {
+		fs.now = opts.Clock.Now
+	} else {
+		fs.now = time.Now
+	}
+	return fs
+}
+
+type compositeFileSystem struct {
+	logf         logger.Logf
+	statChildren bool
+	now          func() time.Time
+
+	// childrenMu guards children
+	childrenMu sync.Mutex
+	children   childrenByName
+}
+
+func (cfs *compositeFileSystem) AddChild(child *Child) {
+	cfs.childrenMu.Lock()
+	var oldChild *Child
+	existingChildPosition, existingChildFound := cfs.searchForChild(child.Name)
+	if existingChildFound {
+		oldChild = cfs.children[existingChildPosition]
+		cfs.children[existingChildPosition] = child
+	} else {
+		cfs.children = append(cfs.children, child)
+		sort.Sort(cfs.children)
+	}
+	cfs.childrenMu.Unlock()
+
+	if oldChild != nil {
+		closer, ok := oldChild.FS.(io.Closer)
+		if ok {
+			err := closer.Close()
+			if err != nil {
+				cfs.logf("failed to close child filesystem %v: %v", child.Name, err)
+			}
+		}
+	}
+}
+
+func (cfs *compositeFileSystem) RemoveChild(name string) {
+	cfs.childrenMu.Lock()
```

Comment:
I think it's simpler to `defer cfs.childrenMu.Unlock()` here. The only thing we're doing outside the lock is (maybe) calling a close function, and we are not shaving cycles here given how much other allocation and sorting we're doing. (Plus, these operations are pretty uncommon, not inner loop stuff)

Reply:
I'm not really worried about the overhead, I'm worried about delays. I've been burned in the past where a Closer's `Close()` actually doesn't return immediately, for example because it's waiting for some work to finish, so I don't want to take the chance of locking the mutex here.